README.md
Wikimedia ObjectFactory
=======================

Construct objects from configuration instructions.

It can be used statically, or as a service wrapping a PSR-11 service container
for lazy instantiation of objects with dependency injection.

Specification array
-------------------

Contents of the specification array are as follows:

    'factory' => callable,
    'class' => string,

The specification array must contain either a 'class' key with string value
that specifies the class name to instantiate or a 'factory' key with a
callable (is_callable() === true). If both are passed, 'factory' takes
precedence but an InvalidArgumentException will be thrown if the resulting
object is not an instance of the named class.

    'args' => array,
    'closure_expansion' => bool, // default true
    'spec_is_arg' => bool, // default false
    'services' => (string|null)[], // default empty

The 'args' key, if provided, specifies arguments to pass to the constructor/callable.
Values in 'args' which are Closure instances will be expanded by invoking
them with no arguments before passing the resulting value on to the
constructor/callable. This can be used to pass live objects to the
constructor/callable. This behavior can be suppressed by adding
closure_expansion => false to the specification.

If 'spec_is_arg' => true is in the specification, 'args' is ignored. The
entire spec array is passed to the constructor/callable instead.

If 'services' is supplied and non-empty (and a service container is available),
the named services are requested from the PSR-11 service container and
prepended before 'args'. `null` values in 'services' are passed to the constructor
unchanged.

If any extra arguments are passed in the options to getObjectFromSpec() or
createObject(), these are prepended before the 'services' and 'args'.
These are passed using the `extraArgs` key, and any Closure instances that
are included this way are never expanded, regardless of whether the main
argument list is expanded.

    'calls' => array

The specification may also contain a 'calls' key that describes method
calls to make on the newly created object before returning it. This
pattern is often known as "setter injection". The value of this key is
expected to be an associative array with method names as keys and
argument lists as values. The argument list will be expanded (or not)
in the same way as the 'args' key for the main object.

Note these calls are not passed the extra arguments.

Installation
------------

```
$ composer require wikimedia/object-factory
```

Usage
-----

```php
<?php

$specs = [
	// Simple constructor based injection
	'testDB' => [
		'class' => PDO::class,
			'args' => [
				'mysql:dbname=testdb;host=127.0.0.1',
				'dbuser',
				'dbpass',
			],
	],
];

$db = ObjectFactory::getObjectFromSpec( $specs['testDB'] ):
```

License
-------
Wikimedia ObjectFactory is licensed under the GNU General Public License,
version 2 and any later version (GPL-2.0-or-later). See the
[`COPYING`](COPYING) file for more details.


HISTORY.md
# Release History

## v5.0.0
* BREAKING CHANGE: drop PHP 7.2/7.3 support (Reedy)
* BREAKING CHANGE: drop `specIsArg` option (DannyS712)
* Updates now we can use PHP 7.4 (Reedy)
* composer.json: Allow psr/container 2 (Reedy)

## v4.0.0
* BREAKING CHANGE: Drop Wikimedia\ObjectFactory class alias (Reedy)

## v3.0.2
* Fix PSR-4 namespace loading issues (Reedy)

## v3.0.1
* Move under new Wikimedia\ObjectFactory namespace (DannyS712)
* Bump psr/container from 1.0.0 to 1.1.1 (Reedy)

## v3.0.0
* BREAKING CHANGE: Drop PHP 7.0/7.1 and HHVM support (James D. Forrester)
* BREAKING CHANGE: Remove deprecated ObjectFactory::constructClassInstance method (DannyS712)
* Allow null values for 'services' spec field (Gergő Tisza)
* Add support for optional services (DannyS712)
* Add phan annotations for better analysis (Daimona Eaytoy)

## v2.1.0
* Implement spec 'services' for DI (Brad Jorsch)

## v2.0.0
* BREAKING CHANGE: Bump required PHP version to 5.6.99 (PHP 7 or HHVM) (Reedy)
* Enhance to be able to replace most/all MediaWiki object instantiation (Brad Jorsch)
* Deprecate ObjectFactory::constructClassInstance method (Brad Jorsch)

## v1.0.0
* Initial commit (Bryan Davis, Kunal Mehta)


tests/ObjectFactoryTest.php
<?php
/**
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 * http://www.gnu.org/copyleft/gpl.html
 *
 * @file
 */

namespace Wikimedia\ObjectFactory\Test;

use Closure;
use Exception;
use InvalidArgumentException;
use PHPUnit\Framework\TestCase;
use Psr\Container\ContainerInterface;
use UnexpectedValueException;
use Wikimedia\ObjectFactory\ObjectFactory;

/**
 * @covers \Wikimedia\ObjectFactory\ObjectFactory
 */
class ObjectFactoryTest extends TestCase {

	public function testClosureExpansionDisabled() {
		$obj = ObjectFactory::getObjectFromSpec( [
			'class' => ObjectFactoryTestFixture::class,
			'args' => [
				static function () {
					return 'wrapped';
				},
				'unwrapped',
			],
			'calls' => [
				'setter' => [ static function () {
					return 'wrapped';
				}, ],
			],
			'closure_expansion' => false,
		] );
		$this->assertInstanceOf( Closure::class, $obj->args[0] );
		$this->assertSame( 'wrapped', $obj->args[0]() );
		$this->assertSame( 'unwrapped', $obj->args[1] );
		$this->assertInstanceOf( Closure::class, $obj->setterArgs[0] );
		$this->assertSame( 'wrapped', $obj->setterArgs[0]() );
	}

	public function testClosureExpansionEnabled() {
		$obj = ObjectFactory::getObjectFromSpec( [
			'class' => ObjectFactoryTestFixture::class,
			'args' => [
				static function () {
					return 'wrapped';
				},
				'unwrapped',
			],
			'calls' => [
				'setter' => [ static function () {
					return 'wrapped';
				}, ],
			],
			'closure_expansion' => true,
		] );
		$this->assertIsString( $obj->args[0] );
		$this->assertSame( 'wrapped', $obj->args[0] );
		$this->assertSame( 'unwrapped', $obj->args[1] );
		$this->assertIsString( $obj->setterArgs[0] );
		$this->assertSame( 'wrapped', $obj->setterArgs[0] );

		$obj = ObjectFactory::getObjectFromSpec( [
			'class' => ObjectFactoryTestFixture::class,
			'args' => [ static function () {
				return 'unwrapped';
			}, ],
			'calls' => [
				'setter' => [ static function () {
					return 'unwrapped';
				}, ],
			],
		] );
		$this->assertIsString( $obj->args[0] );
		$this->assertSame( 'unwrapped', $obj->args[0] );
		$this->assertIsString( $obj->setterArgs[0] );
		$this->assertSame( 'unwrapped', $obj->setterArgs[0] );
	}

	public function testSpecIsArg() {
		$spec = [
			'class' => ObjectFactoryTestFixture::class,
			'args' => [ 'a', 'b' ],
			'spec_is_arg' => true,
		];
		$obj = ObjectFactory::getObjectFromSpec( $spec );
		$this->assertSame( [ $spec ], $obj->args );

		// spec_is_arg defaults to false
		$opts = [ 'extraArgs' => [ 'foo', 'bar' ] ];
		$spec = [
			'class' => ObjectFactoryTestFixture::class,
			'args' => [ 'a', 'b' ],
		];
		$obj = ObjectFactory::getObjectFromSpec( $spec, $opts );
		$this->assertSame( [ 'foo', 'bar', 'a', 'b' ], $obj->args );

		$spec = [
			'class' => ObjectFactoryTestFixture::class,
			'args' => [ 'a', 'b' ],
			'spec_is_arg' => false,
		];
		$obj = ObjectFactory::getObjectFromSpec( $spec, $opts );
		$this->assertSame( [ 'foo', 'bar', 'a', 'b' ], $obj->args );
	}

	public function testGetObjectFromFactory() {
		$args = [ 'a', 'b' ];
		$obj = ObjectFactory::getObjectFromSpec( [
			'factory' => static function ( $a, $b ) {
				return new ObjectFactoryTestFixture( $a, $b );
			},
			'args' => $args,
		] );
		$this->assertSame( $args, $obj->args );
	}

	public function testGetObjectFromInvalid() {
		$args = [ 'a', 'b' ];
		$this->expectException( InvalidArgumentException::class );
		$this->expectExceptionMessage(
			'Provided specification lacks both \'factory\' and \'class\' parameters.'
		);
		ObjectFactory::getObjectFromSpec( [
			// Missing 'class' or 'factory'
			'args' => $args,
		] );
	}

	/**
	 * @dataProvider provideConstructClassInstance
	 */
	public function testGetObjectFromClass( $args ) {
		$obj = ObjectFactory::getObjectFromSpec( [
			'class' => ObjectFactoryTestFixture::class,
			'args' => $args,
		] );
		$this->assertSame( $args, $obj->args );
	}

	public static function provideConstructClassInstance() {
		// These args go to 11. I thought about making 10 one louder, but 11!
		return [
			'0 args' => [ [] ],
			'1 args' => [ [ 1, ] ],
			'2 args' => [ [ 1, 2, ] ],
			'3 args' => [ [ 1, 2, 3, ] ],
			'4 args' => [ [ 1, 2, 3, 4, ] ],
			'5 args' => [ [ 1, 2, 3, 4, 5, ] ],
			'6 args' => [ [ 1, 2, 3, 4, 5, 6, ] ],
			'7 args' => [ [ 1, 2, 3, 4, 5, 6, 7, ] ],
			'8 args' => [ [ 1, 2, 3, 4, 5, 6, 7, 8, ] ],
			'9 args' => [ [ 1, 2, 3, 4, 5, 6, 7, 8, 9, ] ],
			'10 args' => [ [ 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, ] ],
			'11 args' => [ [ 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, ] ],
		];
	}

	public function testNamedArgs() {
		$args = [ 'foo' => 1, 'bar' => 2, 'baz' => 3 ];
		$this->expectException( InvalidArgumentException::class );
		$this->expectExceptionMessage( '\'args\' cannot be an associative array' );
		ObjectFactory::getObjectFromSpec( [
			'class' => ObjectFactoryTestFixture::class,
			'args' => $args,
		] );
	}

	public function testNonObjectFactory() {
		$this->expectException( UnexpectedValueException::class );
		$this->expectExceptionMessage( '\'factory\' did not return an object' );
		ObjectFactory::getObjectFromSpec( [
			'factory' => static function () {
				return null;
			},
		] );
	}

	public function testWrongObjectFactory() {
		$this->expectException( UnexpectedValueException::class );
		$this->expectExceptionMessage( '\'factory\' was expected to return an instance of' );
		ObjectFactory::getObjectFromSpec( [
			'class' => ObjectFactoryTestFixture::class,
			'factory' => static function () {
				return new \stdClass;
			},
		] );
	}

	public function testExtraArgs() {
		$obj = ObjectFactory::getObjectFromSpec( [
			'class' => ObjectFactoryTestFixture::class,
			'args' => [ 'a', 'b' ],
		], [ 'extraArgs' => [ 'foo', 'bar' ] ] );
		$this->assertSame( [ 'foo', 'bar', 'a', 'b' ], $obj->args );
	}

	public function testAssertClass() {
		// This one passes
		ObjectFactory::getObjectFromSpec( [
			'class' => ObjectFactoryTestFixture::class,
		], [ 'assertClass' => ObjectFactoryTestFixture::class ] );

		// This one fails
		$this->expectException( UnexpectedValueException::class );
		$this->expectExceptionMessage( 'Expected instance of FooBar, got' );
		ObjectFactory::getObjectFromSpec( [
			'class' => ObjectFactoryTestFixture::class,
		], [ 'assertClass' => 'FooBar' ] );
	}

	public function testClassSpecNotAllowed() {
		$this->expectException( InvalidArgumentException::class );
		$this->expectExceptionMessage( 'Passing a raw class name is not allowed here' );
		ObjectFactory::getObjectFromSpec(
			ObjectFactoryTestFixture::class,
			[ 'extraArgs' => [ 'foo', 'bar' ] ]
		);
	}

	public function testClassSpecAllowed() {
		$obj = ObjectFactory::getObjectFromSpec(
			ObjectFactoryTestFixture::class,
			[ 'allowClassName' => true, 'extraArgs' => [ 'foo', 'bar' ] ]
		);
		$this->assertSame( [ 'foo', 'bar' ], $obj->args );
	}

	public function testCallableSpecNotAllowed() {
		$this->expectException( InvalidArgumentException::class );
		$this->expectExceptionMessage( 'Passing a raw callable is not allowed here' );
		ObjectFactory::getObjectFromSpec(
			static function ( ...$args ) {
				return new ObjectFactoryTestFixture( ...$args );
			},
			[ 'extraArgs' => [ 'foo', 'bar' ] ]
		);
	}

	public function testCallableSpecAllowed() {
		$obj = ObjectFactory::getObjectFromSpec(
			static function ( ...$args ) {
				return new ObjectFactoryTestFixture( ...$args );
			},
			[ 'allowCallable' => true, 'extraArgs' => [ 'foo', 'bar' ] ]
		);
		$this->assertSame( [ 'foo', 'bar' ], $obj->args );
	}

	public function testBadSpec() {
		$this->expectException( InvalidArgumentException::class );
		$this->expectExceptionMessage( 'Provided specification is not an array.' );
		ObjectFactory::getObjectFromSpec(
			'ThisDoesNotExist',
			[ 'allowClassName' => true,  'allowCallable' => true ]
		);
	}

	public function testServices() {
		$services = [
			'Foo' => (object)[ 'foo' ],
			'Bar' => (object)[ 'bar' ],
			'Baz' => (object)[ 'baz' ],
		];

		$container = $this->createMock( ContainerInterface::class );
		$container->method( 'get' )->willReturnCallback( static function ( $name ) use ( $services ) {
			if ( isset( $services[$name] ) ) {
				return $services[$name];
			}
			throw new Exception( "Service $name not found" );
		} );
		$container->method( 'has' )->willReturnCallback( static function ( $name ) use ( $services ) {
			return isset( $services[$name] );
		} );

		// Basic usage
		$obj = ObjectFactory::getObjectFromSpec(
			[
				'class' => ObjectFactoryTestFixture::class,
				'services' => [ 'Foo' ],
			],
			[
				'serviceContainer' => $container,
			]
		);
		$this->assertSame( [ $services['Foo'] ], $obj->args );

		// Ordering of argument sources
		$obj = ObjectFactory::getObjectFromSpec(
			[
				'class' => ObjectFactoryTestFixture::class,
				'args' => [ 'a', 'b' ],
				'services' => [ 'Bar', 'Foo' ],
			],
			[
				'extraArgs' => [ 'x', 'y' ],
				'serviceContainer' => $container,
			]
		);
		$this->assertSame( [ 'x', 'y', $services['Bar'], $services['Foo'], 'a', 'b' ], $obj->args );

		// Repetition of a service
		$spec = [
			'factory' => static function ( ...$args ) {
				return new ObjectFactoryTestFixture( ...$args );
			},
			'spec_is_arg' => true,
			'services' => [ 'Baz', 'Baz' ],
		];
		$obj = ObjectFactory::getObjectFromSpec( $spec, [
			'extraArgs' => [ 'x', 'y' ],
			'serviceContainer' => $container,
		] );
		$this->assertSame( [ 'x', 'y', $services['Baz'], $services['Baz'], $spec ], $obj->args );

		// Null service passed as null
		$obj = ObjectFactory::getObjectFromSpec(
			[
				'class' => ObjectFactoryTestFixture::class,
				'services' => [ null, 'Foo', null, 'Bar' ],
			],
			[
				'serviceContainer' => $container,
			]
		);
		$this->assertSame( [ null, $services['Foo'], null, $services['Bar'] ], $obj->args );

		// Optional services
		$obj = ObjectFactory::getObjectFromSpec(
			[
				'class' => ObjectFactoryTestFixture::class,
				'optional_services' => [ 'ServiceThatDoesNotExist', 'Foo' ]
			],
			[
				'serviceContainer' => $container
			]
		);
		$this->assertSame( [ null, $services['Foo'] ], $obj->args );

		// Order of arguments with optional and non-optional services
		$obj = ObjectFactory::getObjectFromSpec(
			[
				'class' => ObjectFactoryTestFixture::class,
				'args' => [ 'param1', 'param2' ],
				'services' => [ 'Bar', 'Foo' ],
				'optional_services' => [ 'ServiceThatDoesNotExist', 'Foo' ],
			],
			[
				'extraArgs' => [ 'extra1', 'extra2' ],
				'serviceContainer' => $container,
			]
		);
		$this->assertSame(
			[
				'extra1',
				'extra2',
				$services['Bar'],
				$services['Foo'],
				null,
				$services['Foo'],
				'param1',
				'param2'
			],
			$obj->args
		);
	}

	/**
	 * @dataProvider provideTestMissingServiceContainer
	 * @param string $type either 'services' or 'optional_services'
	 */
	public function testServices_error( $type ) {
		$this->expectException( InvalidArgumentException::class );
		$this->expectExceptionMessage(
			'\'services\' and \'optional_services\' cannot be used without a service container'
		);

		$spec = [ 'class' => ObjectFactoryTestFixture::class ];
		$spec[ $type ] = [ 'foo', 'bar' ];
		ObjectFactory::getObjectFromSpec( $spec );
	}

	public function provideTestMissingServiceContainer() {
		return [
			'Normal required services' => [ 'services' ],
			'Optional services' => [ 'optional_services' ],
		];
	}

	public function testNonStaticUse() {
		$container = $this->getMockBuilder( ContainerInterface::class )
			->getMockForAbstractClass();

		// Can't mock a static method, but we can make an anonymous subclass overriding it.
		$factory = new class ( $container ) extends ObjectFactory {
			public static function getObjectFromSpec( $spec, array $options = [] ) {
				$that = $options['that'];
				$that->assertArrayHasKey( 'serviceContainer', $options );
				$that->assertInstanceOf( ContainerInterface::class, $options['serviceContainer'] );
				return parent::getObjectFromSpec( $spec, $options );
			}
		};

		$obj = $factory->createObject( [
			'class' => ObjectFactoryTestFixture::class,
			'args' => [ 'a', 'b' ],
		], [ 'that' => $this ] );
		$this->assertSame( [ 'a', 'b' ], $obj->args );
	}

}


tests/ObjectFactoryTestFixture.php
<?php
/**
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 * http://www.gnu.org/copyleft/gpl.html
 *
 * @file
 */

namespace Wikimedia\ObjectFactory\Test;

class ObjectFactoryTestFixture {
	/**
	 * @var mixed Arguments passed to constructor
	 */
	public $args;

	/**
	 * @var mixed Arguments passed to setter
	 */
	public $setterArgs;

	public function __construct() {
		$this->args = func_get_args();
	}

	/**
	 * Dependency injection setter stub.
	 */
	public function setter() {
		$this->setterArgs = func_get_args();
	}
}


src/ObjectFactory.php
<?php
/**
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 * http://www.gnu.org/copyleft/gpl.html
 *
 * @file
 */

namespace Wikimedia\ObjectFactory;

use Closure;
use InvalidArgumentException;
use Psr\Container\ContainerInterface;
use UnexpectedValueException;

/**
 * Construct objects based on a specification array.
 *
 * Contents of the specification array are as follows:
 *
 *     'factory' => callable,
 *     'class' => string,
 *
 * The specification array must contain either a 'class' key with string value
 * that specifies the class name to instantiate or a 'factory' key with a
 * callable (is_callable() === true). If both are passed, 'factory' takes
 * precedence but an InvalidArgumentException will be thrown if the resulting
 * object is not an instance of the named class.
 *
 *     'args' => array,
 *     'closure_expansion' => bool, // default true
 *     'spec_is_arg' => bool, // default false
 *     'services' => (string|null)[], // default empty
 *     'optional_services' => (string|null)[], // default empty
 *
 * The 'args' key, if provided, specifies arguments to pass to the constructor/callable.
 * Values in 'args' which are Closure instances will be expanded by invoking
 * them with no arguments before passing the resulting value on to the
 * constructor/callable. This can be used to pass live objects to the
 * constructor/callable. This behavior can be suppressed by adding
 * closure_expansion => false to the specification.
 *
 * If 'spec_is_arg' => true is in the specification, 'args' is ignored. The
 * entire spec array is passed to the constructor/callable instead.
 *
 * If 'services' is supplied and non-empty (and a service container is available),
 * the named services are requested from the PSR-11 service container and
 * prepended before 'args'. `null` values in 'services' are passed to the constructor
 * unchanged.
 *
 * Optional services declared via 'optional_services' are handled the same,
 * except that if the service is not available from the service container
 * `null` is passed as a parameter instead. Optional services are appended
 * directly after the normal required services.
 *
 * If any extra arguments are passed in the options to getObjectFromSpec() or
 * createObject(), these are prepended before the 'services' and 'args'.
 *
 *     'calls' => array
 *
 * The specification may also contain a 'calls' key that describes method
 * calls to make on the newly created object before returning it. This
 * pattern is often known as "setter injection". The value of this key is
 * expected to be an associative array with method names as keys and
 * argument lists as values. The argument list will be expanded (or not)
 * in the same way as the 'args' key for the main object.
 *
 * Note these calls are not passed the extra arguments.
 *
 * @copyright © 2014 Wikimedia Foundation and contributors
 */
class ObjectFactory {

	/** @var ContainerInterface Service container */
	protected $serviceContainer;

	/**
	 * @param ContainerInterface $serviceContainer Service container
	 */
	public function __construct( ContainerInterface $serviceContainer ) {
		$this->serviceContainer = $serviceContainer;
	}

	/**
	 * Instantiate an object based on a specification array.
	 *
	 * This calls getObjectFromSpec(), with the ContainerInterface that was
	 * passed to the constructor passed as `$options['serviceContainer']`.
	 *
	 * @phan-template T
	 * @phpcs:disable Generic.Files.LineLength
	 * @phan-param class-string<T>|callable(mixed ...$args):T|array{class?:class-string<T>,factory?:callable(mixed ...$args):T,args?:array,services?:array<string|null>,optional_services?:array<string|null>,calls?:string[],closure_expansion?:bool,spec_is_arg?:bool} $spec
	 * @phan-param array{allowClassName?:bool,allowCallable?:bool,extraArgs?:array,assertClass?:string} $options
	 * @phpcs:enable
	 * @phan-return T|object
	 *
	 * @param array|string|callable $spec Specification array, or (when the respective
	 *   $options flag is set) a class name or callable. Allowed fields (see class
	 *   documentation for more details):
	 *   - 'class': (string) Class of the object to create. If 'factory' is also specified,
	 *     it will be used to validate the object.
	 *   - 'factory': (callable) Factory method for creating the object.
	 *   - 'args': (array) Arguments to pass to the constructor or the factory method.
	 *   - 'services': (array of string/null) List of services to pass as arguments. Each
	 *     name will be looked up in the container given to ObjectFactory in its constructor,
	 *     and the results prepended to the argument list. Null values are passed unchanged.
	 *   - 'optional_services': (array of string/null) Handled the same as services, but if
	 *     the service is unavailable from the service container the parameter is set to 'null'
	 *     instead of causing an error.
	 *   - 'calls': (array) A list of calls to perform on the created object, for setter
	 *     injection. Keys of the array are method names and values are argument lists
	 *     (as arrays). These arguments are not affected by any of the other specification
	 *     fields that manipulate constructor arguments.
	 *   - 'closure_expansion': (bool, default true) Whether to expand (execute) closures
	 *     in 'args'.
	 *   - 'spec_is_arg': (bool, default false) When true, 'args' is ignored and the entire
	 *     specification array is passed as an argument.
	 *   One of 'class' and 'factory' is required.
	 * @param array $options Allowed keys are
	 *  - 'allowClassName': (bool) If set and truthy, $spec may be a string class name.
	 *    In this case, it will be treated as if it were `[ 'class' => $spec ]`.
	 *  - 'allowCallable': (bool) If set and truthy, $spec may be a callable. In this
	 *    case, it will be treated as if it were `[ 'factory' => $spec ]`.
	 *  - 'extraArgs': (array) Extra arguments to pass to the constructor/callable. These
	 *    will come before services and normal args.
	 *  - 'assertClass': (string) Throw an UnexpectedValueException if the spec
	 *    does not create an object of this class.
	 * @return object
	 * @throws InvalidArgumentException when object specification is not valid.
	 * @throws UnexpectedValueException when the factory returns a non-object, or
	 *  the object is not an instance of the specified class.
	 */
	public function createObject( $spec, array $options = [] ) {
		$options['serviceContainer'] = $this->serviceContainer;
		// ObjectFactory::getObjectFromSpec accepts an array, not just a callable (phan bug)
		// @phan-suppress-next-line PhanTypeInvalidCallableArraySize
		return static::getObjectFromSpec( $spec, $options );
	}

	/**
	 * Instantiate an object based on a specification array.
	 *
	 * @phan-template T
	 * @phpcs:disable Generic.Files.LineLength
	 * @phan-param class-string<T>|callable(mixed ...$args):T|array{class?:class-string<T>,factory?:callable(mixed ...$args):T,args?:array,services?:array<string|null>,optional_services?:array<string|null>,calls?:string[],closure_expansion?:bool,spec_is_arg?:bool} $spec
	 * @phan-param array{allowClassName?:bool,allowCallable?:bool,extraArgs?:array,assertClass?:string,serviceContainer?:ContainerInterface} $options
	 * @phpcs:enable
	 * @phan-return T|object
	 *
	 * @param array|string|callable $spec As for createObject().
	 * @param array $options As for createObject(). Additionally:
	 *  - 'serviceContainer': (ContainerInterface) PSR-11 service container to use
	 *    to handle 'services'.
	 * @return object
	 * @throws InvalidArgumentException when object specification is not valid.
	 * @throws InvalidArgumentException when $spec['services'] or $spec['optional_services']
	 *  is used without $options['serviceContainer'] being set and implementing ContainerInterface.
	 * @throws UnexpectedValueException when the factory returns a non-object, or
	 *  the object is not an instance of the specified class.
	 */
	public static function getObjectFromSpec( $spec, array $options = [] ) {
		$spec = static::validateSpec( $spec, $options );

		$expandArgs = !isset( $spec['closure_expansion'] ) || $spec['closure_expansion'];

		if ( !empty( $spec['spec_is_arg'] ) ) {
			$args = [ $spec ];
		} else {
			$args = $spec['args'] ?? [];

			// $args should be a non-associative array; show nice error if that's not the case
			if ( $args && array_keys( $args ) !== range( 0, count( $args ) - 1 ) ) {
				throw new InvalidArgumentException( '\'args\' cannot be an associative array' );
			}

			if ( $expandArgs ) {
				$args = static::expandClosures( $args );
			}
		}

		$services = [];
		if ( !empty( $spec['services'] ) || !empty( $spec['optional_services'] ) ) {
			$container = $options['serviceContainer'] ?? null;
			if ( !$container instanceof ContainerInterface ) {
				throw new InvalidArgumentException(
					'\'services\' and \'optional_services\' cannot be used without a service container'
				);
			}

			if ( !empty( $spec['services'] ) ) {
				foreach ( $spec['services'] as $service ) {
					$services[] = $service === null ? null : $container->get( $service );
				}
			}

			if ( !empty( $spec['optional_services'] ) ) {
				foreach ( $spec['optional_services'] as $service ) {
					if ( $service !== null && $container->has( $service ) ) {
						$services[] = $container->get( $service );
					} else {
						// Either $service was null, or the service was not available
						$services[] = null;
					}
				}
			}
		}

		$args = array_merge(
			$options['extraArgs'] ?? [],
			$services,
			$args
		);

		if ( isset( $spec['factory'] ) ) {
			$obj = $spec['factory']( ...$args );
			if ( !is_object( $obj ) ) {
				throw new UnexpectedValueException( '\'factory\' did not return an object' );
			}
			// @phan-suppress-next-line PhanRedundantCondition
			if ( isset( $spec['class'] ) && !$obj instanceof $spec['class'] ) {
				throw new UnexpectedValueException(
					'\'factory\' was expected to return an instance of ' . $spec['class']
					. ', got ' . get_class( $obj )
				);
			}
		} elseif ( isset( $spec['class'] ) ) {
			$clazz = $spec['class'];
			$obj = new $clazz( ...$args );
		} else {
			throw new InvalidArgumentException(
				'Provided specification lacks both \'factory\' and \'class\' parameters.'
			);
		}

		// @phan-suppress-next-line PhanRedundantCondition
		if ( isset( $options['assertClass'] ) && !$obj instanceof $options['assertClass'] ) {
			throw new UnexpectedValueException(
				'Expected instance of ' . $options['assertClass'] . ', got ' . get_class( $obj )
			);
		}

		if ( isset( $spec['calls'] ) && is_array( $spec['calls'] ) ) {
			// Call additional methods on the newly created object
			foreach ( $spec['calls'] as $method => $margs ) {
				if ( $expandArgs ) {
					$margs = static::expandClosures( $margs );
				}
				call_user_func_array( [ $obj, $method ], $margs );
			}
		}

		return $obj;
	}

	/**
	 * Convert a string or callable to a spec array
	 *
	 * @param array|string|callable $spec As for createObject() or getObjectFromSpec()
	 * @param array $options As for createObject() or getObjectFromSpec()
	 * @return array Specification array
	 * @throws InvalidArgumentException when object specification does not
	 *  contain 'class' or 'factory' keys
	 */
	protected static function validateSpec( $spec, array $options ) {
		if ( is_callable( $spec ) ) {
			if ( empty( $options['allowCallable'] ) ) {
				throw new InvalidArgumentException(
					'Passing a raw callable is not allowed here. Use [ \'factory\' => $callable ] instead.'
				);
			}
			return [ 'factory' => $spec ];
		}
		if ( is_string( $spec ) && class_exists( $spec ) ) {
			if ( empty( $options['allowClassName'] ) ) {
				throw new InvalidArgumentException(
					'Passing a raw class name is not allowed here. Use [ \'class\' => $classname ] instead.'
				);
			}
			return [ 'class' => $spec ];
		}

		if ( !is_array( $spec ) ) {
			throw new InvalidArgumentException( 'Provided specification is not an array.' );
		}

		return $spec;
	}

	/**
	 * Iterate a list and call any closures it contains.
	 *
	 * @param array $list List of things
	 * @return array List with any Closures replaced with their output
	 */
	protected static function expandClosures( $list ) {
		return array_map( static function ( $value ) {
			if ( is_object( $value ) && $value instanceof Closure ) {
				// If $value is a Closure, call it.
				return $value();
			} else {
				return $value;
			}
		}, $list );
	}

}
